engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/number/bound/MaxValidatorForNumber.java

```diff
@@ -16,6 +18,13 @@
 
 	@Override
 	protected int compare(Number number) {
+		// At this point number can be a floating point number and it should be treated differently in such case:
+		if ( number instanceof Double ) {
+			return NumberComparatorHelper.compare( (Double) number, maxValue, InfinityNumberComparatorHelper.GREATER_THAN );
+		}
+		if ( number instanceof Float ) {
+			return NumberComparatorHelper.compare( (Float) number, maxValue, InfinityNumberComparatorHelper.GREATER_THAN );
+		}
```

Comment:
it's because of that third parameter. Initially I've put it all in `NumberComparatorHelper`. But we need to know how to treat infinities. So it was either this or add third parameter to all compare methods from `NumberComparatorHelper`.

Reply:
Maybe not all methods but it would make sense to have it for the `Number` one and push the logic there.